### PR TITLE
Remove channel that sends roots to BlockstoreCleanupService

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
Currently, ReplayStage sends new roots to BlockstoreCleanupService, and
BlockstoreCleanupService decides when to clean based on the advancement
of the latest root. This is totally unnecessary as the latest root is
cached by the Blockstore, and this value can simply be fetched.

#### Summary of Changes
- Rip out the channel, and instead query latest root from Blockstore directly
- Eliminate separate thread that was spawned to do the cleanup while the channel was drained
- Remove `latest_root_senders` from `ReplayStage` as the only receiver of these roots was `BlockstoreCleanupService`

#### Impact
Prior to this change, we were sending / receiving up to 150 roots/min (assuming 0% skip rate) across the channel. With the change, we are now reading an atomic every (512 * 0.4 / 10) ~20 seconds instead of using the channel.

Arguably, we could possibly increase the ~20s value to a larger interval (ie read less frequently) with the tradeoff that we may not clean as quickly as we could. However, I think there is minimal benefit to further tuning this value; the use of `LOOP_LIMITER` already makes us read the atomic 1/20 as often as we would if the `LOOP_LIMITER` if-statement didn't exist, and we just checked after every `1s` sleep

### CODE REVIEW

**Comment by steviez** [comment]
Ok, I think this is fully good for review. I've been running it for ~24 hours and the node is still performing cleanup as expected, on a regular interval.

Quick comments on `LOOP_LIMITER`:
- `DEFAULT_CLEANUP_SLOT_INTERVAL == 512` means there will be cleanup to do roughly every `512 * 0.4 == ~200` seconds
- I initially had `LOOP_LIMITER` set to half of this value, or roughly 100 seconds. What I observed is that my node would enter cleanup after `LOOP_LIMITER` had elapsed 3 times, or about every 300 seconds.

So, there was technically cleanup to do every 200 seconds, but it was only getting done every 300 seconds. This lead to me changing the `/ 2` to `/ 10` in `LOOP_LIMITER`. This shrinks the gap between when there is cleanup to do and when we pick it up. Cleanup in general is pretty coarse, so any variance here is just further noise in the variations we might see from roots not advancing (ie forking) or variability in block sizes (ie high / low usage)

File: ledger/src/blockstore_cleanup_service.rs [hunk 1]
```diff
@@ -36,46 +35,53 @@ pub const DEFAULT_MAX_LEDGER_SHREDS: u64 = 200_000_000;
...
 pub const DEFAULT_MIN_MAX_LEDGER_SHREDS: u64 = 50_000_000;
 
-// Check for removing slots at this interval so we don't purge too often
-// and starve other blockstore users.
-pub const DEFAULT_PURGE_SLOT_INTERVAL: u64 = 512;
+// Perform blockstore cleanup at this interval to limit the overhead of cleanup
+// Cleanup will be considered after the latest root has advanced by this value
+const DEFAULT_CLEANUP_SLOT_INTERVAL: u64 = 512;
+// The above slot interval can be roughly equated to a time interval. So, scale
+// how often we check for cleanup with the interval. Doing so will avoid easted
```
Comment (AshwinSekar) [inline ledger/src/blockstore_cleanup_service.rs:42]:
```suggestion
// how often we check for cleanup with the interval. Doing so will avoid wasted
```

**Review by AshwinSekar — COMMENTED** [review]
just a couple of nits, logic looks good to me. I think `/ 10` is fine for `LOOP_LIMITER`, as you pointed out cleanup is pretty coarse so delaying by a handful of slots should be fine.

🪓 the redundant channels.

File: ledger/src/blockstore_cleanup_service.rs [hunk 2]
```diff
@@ -169,107 +175,66 @@ impl BlockstoreCleanupService {
...
-        let root = Self::receive_new_roots(new_root_receiver)?;
+    ) {
+        let root = blockstore.max_root();
         if root - *last_purge_slot <= purge_interval {
-            return Ok(());
+            return;
         }
-
-        let disk_utilization_pre = blockstore.storage_size();
-        info!(
```
[the remark this answers is not in the document]
Comment (steviez) [reply ledger/src/blockstore_cleanup_service.rs:213]: > from a functional standpoint it seems that this pre purge log can be deleted, as the post purge log and Self::report_disk_metrics should cover the information.

Exactly, we already report the `pre` value at the end so this was my thinking / justification for removing this first info.

> however it seems we no longer have any indication that a cleanup has been started, in the case that some purge operation hangs. would it be useful to keep this log around for debugging?

Good instincts; I had a similar thought, especially with the `lowest_cleanup_slot` write lock getting grabbed. However, there is an info that gets emitted by `Self::find_slots_to_clean()`:
https://github.com/solana-labs/solana/blob/30adda4a71bf4a63103badbe34f596272b6d716f/ledger/src/blockstore_cleanup_service.rs#L148-L151

So, I was initially inclined to say that one was good. But, given that this service only does stuff every ~3.5 min, I think adding one back in once we know that the latest root has advanced far enough is fine. I did already condense the `purging data older than ...` and `purge_time` info's so I'm still shaving off one `info` with this PR haha

**Review by AshwinSekar — APPROVED** [review]
LGTM thanks for the cleanup!

**Pull request merged**